bittide/src/Bittide/Transceiver/Prbs.hs

```diff
@@ -0,0 +1,137 @@
+-- (@polyLength@), and the second tap is the @polyTap@-th bit.
+--
+-- See https://en.wikipedia.org/wiki/Pseudorandom_binary_sequence for more information.
+data Config polyLength polyTap nBits where
+  Config ::
+    ( KnownNat polyLength
+    , KnownNat polyTap
+    , KnownNat nBits
+
+    , 1 <= nBits
+    , 1 <= polyTap
+    , (polyTap + 1) <= polyLength
+
+    -- Same constraints, but written differently for type checking purposes:
+    , (_n0 + 1) ~ nBits
+    , (polyTap + _n1) ~ polyLength
+    , polyTap ~ (_n2 + 1)
+    , _n1 ~ (_n3 + 1)
+    ) =>
+    Config polyLength polyTap nBits
+
+
+-- | PRBS31: @x^31 + x^28 + 1@
+conf31 :: forall n . (31 <= n, KnownNat n) => Config 31 28 n
+conf31 = leToPlus @31 @n Config
+
+-- | PRBS generator, see module documentation.
+generator ::
+  forall dom polyLength polyTap nBits .
+  KnownDomain dom =>
+  Clock dom -> Reset dom -> Enable dom ->
+  Config polyLength polyTap nBits ->
+  Signal dom (BitVector nBits)
+generator clk rst ena Config =
+  moore clk rst ena go snd (maxBound, maxBound) (pure ())
+ where
+  go ::
+    (BitVector polyLength, BitVector nBits) ->
+    () ->
+    (BitVector polyLength, BitVector nBits)
+  go (prbs_reg, _) _ =
+    ( last prbs
+    , pack (reverse $ map msb prbs) )
+   where
+     prbs :: Vec nBits (BitVector polyLength)
+     prbs = unfoldrI goPrbs prbs_reg
+
+     goPrbs :: BitVector polyLength -> (BitVector polyLength, BitVector polyLength)
+     goPrbs bv = (o,o)
+      where
+       o = newBit +>>. bv
+       tap = SNat @(polyLength - polyTap)
+       newBit = xor (lsb bv) (unpack $ slice tap tap bv)
+
+
+-- | PRBS checker, see module documentation.
+checker ::
+  forall dom polyLength polyTap nBits .
+  KnownDomain dom =>
+  Clock dom -> Reset dom -> Enable dom ->
+  Config polyLength polyTap nBits ->
+  Signal dom (BitVector nBits) ->
+  Signal dom (BitVector nBits)
+checker clk rst ena Config = mealy clk rst ena go (maxBound, maxBound)
+ where
+  go ::
+    (BitVector polyLength, BitVector nBits) ->
+    BitVector nBits ->
+    ((BitVector polyLength, BitVector nBits), BitVector nBits)
+  go (prbs_reg, prbs_out_prev) prbsIn =
+    ( (prbs_state, pack $ reverse prbs_out)
+    , prbs_out_prev
+    )
+   where
+     prbs_out :: Vec nBits Bit
+     prbs_state :: BitVector polyLength
+     (prbs_state, prbs_out) = mapAccumL goPrbs prbs_reg (reverse $ unpack prbsIn)
+
+     goPrbs :: BitVector polyLength -> Bit -> (BitVector polyLength, Bit)
+     goPrbs bv newBit = (o, bitErr)
+      where
+       o = newBit +>>. bv
+       tap = SNat @(polyLength - polyTap)
+       bitErr = xor newBit (xor (lsb bv) (unpack $ slice tap tap bv))
+
+data TrackerState
+  = Down (Index 8096)
+  -- ^ Link is considered down. Needs 127 cycles of \"good\" input to transition
+  -- to 'Up'.
+  | Up
+  -- ^ Link has not seen errors in at least 127 cycles.
+  deriving (Eq, Show, Generic, NFDataX)
+
+isUp :: TrackerState -> Bool
+isUp Up = True
+isUp _ = False
```

Comment:
`isUp` is never used.

Reply:
Removed

Reply:
And there it is again :upside_down_face:.

Reply:
The heck :see_no_evil: